src/start/RealmScreen.js

```diff
@@ -53,6 +54,26 @@ class RealmScreen extends PureComponent<Props, State> {
 
     const { dispatch } = this.props;
 
+    let parsedRealm;
+    try {
+      parsedRealm = new WhatwgURL(realm);
+    } catch (err) {
+      this.setState({
+        progress: false,
+        error: 'Please enter a valid URL',
+      });
+      return;
+    }
+
+    if (parsedRealm.username !== '') {
+      this.setState({
+        progress: false,
+        error:
+          'Please enter the server URL, not your email. You can enter your credentials in the next screen.',
+      });
+      return;
+    }
+
     try {
       const serverSettings: ApiResponseServerSettings = await api.getServerSettings(realm);
       dispatch(realmAdd(realm, new ZulipVersion(serverSettings.zulip_version)));
```

Comment:
This function gets somewhat difficult to follow with all the stuff going on in it now. I think one reason is that the `setState` at the top is setting up the start of a story you basically have to read the whole function to see the different endings of.

One thing that would simplify it is to not set `progress: true` until after the realm-parsing is done and those errors haven't happened. That way those don't need to say `progress: false`, and the whole question of the `progress` state is limited to just where we're actually contacting the server, which is what it's for.

Relatedly, we can cut the line already that sets `realm` in that first `setState`. It's a no-op because we just read `realm` *from* the state. Looks like it's a legacy of an unfinished refactor in 252fed22fa1432e4cf6469423099521ecc3ed7ab -- before then, this line actually set `realm` to a different value than it already had, and so was doing something useful.

Similarly to moving the `setState` but smaller: the `const { dispatch } = …` line can wait until just above the `try`.

Reply:
Okay, also, I moved the `error:null` to just above the `try` block too.